src/util/object_pool/object_pool.c

```diff
@@ -46,55 +57,72 @@ static void _ObjectPool_AddBlocks(ObjectPool *pool, uint blockCount) {
 	pool->itemCap = pool->blockCount * POOL_BLOCK_CAP;
 }
 
-ObjectPool *ObjectPool_New(uint itemCap, uint itemSize, fpDestructor fp) {
+// Clear a deleted item and recycle it to the caller.
+static void *_ObjectPool_ReuseItem(ObjectPool *pool) {
+	ObjectID idx = array_pop(pool->deletedIdx);
+
+	pool->itemCount++;
+
+	Block *block = GET_ITEM_BLOCK(pool, idx);
+	uint pos = ITEM_POSITION_WITHIN_BLOCK(idx);
+
+	// Retrieve a pointer to the item's header.
+	unsigned char *item_header = block->data + (pos * block->itemSize);
+	unsigned char *item = ITEM_FROM_HEADER(item_header);
+	assert(ITEM_ID(item) == idx); // The item ID should not change on reuse.
+
+	// Zero-set the item being returned.
+	memset(item, 0, block->itemSize - HEADER_SIZE);
+
+	return item;
+}
+
+ObjectPool *ObjectPool_New(uint64_t itemCap, uint itemSize, fpDestructor fp) {
 	ObjectPool *pool = rm_malloc(sizeof(ObjectPool));
 	pool->itemCount = 0;
-	pool->itemSize = itemSize;
+	pool->itemSize = itemSize + HEADER_SIZE; // Add extra space to accommodate the item's header.
 	pool->blockCount = 0;
 	pool->blocks = NULL;
-	pool->deletedIdx = array_new(uint, 128);
+	pool->deletedIdx = array_new(uint64_t, 128);
 	pool->destructor = fp;
 	_ObjectPool_AddBlocks(pool, ITEM_COUNT_TO_BLOCK_COUNT(itemCap));
 	return pool;
 }
 
-void *ObjectPool_NewItem(ObjectPool *pool, uint *idx) {
+void *ObjectPool_NewItem(ObjectPool *pool) {
+	// Reuse a deleted item if one is available.
+	if(array_len(pool->deletedIdx)) return _ObjectPool_ReuseItem(pool);
+
 	// Make sure we have room for a new item.
 	if(pool->itemCount >= pool->itemCap) {
 		// Double the capacity of the pool.
-		_ObjectPool_AddBlocks(pool, pool->itemCap);
+		_ObjectPool_AddBlocks(pool, ITEM_COUNT_TO_BLOCK_COUNT(pool->itemCap));
 	}
 
-	// Get the index of the allocation, reusing a deleted index if possible.
-	uint pos = (array_len(pool->deletedIdx)) ? array_pop(pool->deletedIdx) : pool->itemCount;
+	// Get the index of the new allocation.
+	ObjectID idx = pool->itemCount;
 	pool->itemCount++;
 
-	*idx = pos;
+	Block *block = GET_ITEM_BLOCK(pool, idx);
+	uint pos = ITEM_POSITION_WITHIN_BLOCK(idx);
 
-	Block *block = GET_ITEM_BLOCK(pool, pos);
-	pos = ITEM_POSITION_WITHIN_BLOCK(pos);
-
-	unsigned char *item = block->data + (pos * block->itemSize);
-	// Zero-set the item being returned.
-	memset(item, 0, block->itemSize);
+	// Retrieve a pointer to the item's header.
+	unsigned char *item_header = block->data + (pos * block->itemSize);
+	unsigned char *item = ITEM_FROM_HEADER(item_header);
+	ITEM_ID(item) = idx; // Set the item ID.
```

Comment:
I thought `ITEM_ID` returns a const, if that's the case I'm not sure how does this update works.